src/components/code-block/index.js

```diff
@@ -0,0 +1,102 @@
+import { useState, useMemo, useRef, useEffect } from 'preact/hooks';
+import { Link } from 'preact-router';
+import cx from '../../lib/cx';
+import PrismWorker from 'workerize-loader?name=prism.[hash:5]!./prism.worker';
+
+const { highlight } = PRERENDER ? require('./prism.worker') : new PrismWorker();
+
+function useFuture(initializer, params) {
+	const getInitialState = () => {
+		try {
+			const value = initializer();
+			if (value && value.then) {
+				if ('_value' in value) return [value._value];
+				if ('_error' in value) return [undefined, value._error];
+				return [undefined, undefined, value];
+			}
+			return [value];
+		} catch (err) {
+			return [undefined, err];
+		}
+	};
+
+	const [pair, setValue] = useState(getInitialState);
+
+	// only run on changes, not initial mount
+	const isFirstRun = useRef(true);
+	useEffect(() => {
+		if (isFirstRun.current) return (isFirstRun.current = false);
+		setValue(getInitialState());
+	}, params || []);
+
+	const pending = pair[2];
+	if (pending) {
+		if (!pending._processing) {
+			pending._processing = true;
+			pending
+				.then(value => {
+					pending._value = value;
+					setValue([value]);
+				})
+				.catch(err => {
+					pending._error = err;
+					setValue([undefined, err]);
+				});
+		}
+	}
+	return pair;
+}
+
+const CACHE = {};
+function cachedHighlight(code, lang) {
+	const id = lang + '\n' + code;
+	return CACHE[id] || (CACHE[id] = highlight(code, lang));
+}
+
+function HighlightedCodeBlock({ code, lang, ...props }) {
+	const [highlighted, error, pending] = useFuture(
+		() => cachedHighlight(code, lang),
+		[code, lang]
+	);
+	const repl =
+		(lang === 'js' || lang === 'jsx') &&
+		code.split('\n').length > 2 &&
+		props.repl !== 'false';
+	const canHighlight = !!pending || !error;
+	const html =
+		(canHighlight && highlighted) ||
+		code.replace(/</g, '&lt;').replace(/>/g, '&gt;');
+	const htmlObj = useMemo(() => ({ __html: html }), [html]);
```

Comment:
We could theoretically skip the `useMemo` here because we compare the html property during diffing:

https://github.com/preactjs/preact/blob/25b15fb1e352f5dbb863a61b5e911d3f264b6da2/src/diff/index.js#L232

Reply:
I'll check - I think I was seeing reapplication of HTML here, but you're right there is no reason Preact would be doing that. Was likely something else and I can un-patch.